pulsar-broker-common/src/main/java/org/apache/pulsar/broker/authentication/AuthenticationProviderToken.java

```diff
@@ -0,0 +1,147 @@
+
+    // When public/private key pair is configured
+    final static String CONF_TOKEN_PUBLIC_KEY = "tokenPublicKey";
+
+    private Key validationKey;
+
+    @Override
+    public void close() throws IOException {
+        // noop
+    }
+
+    @Override
+    public void initialize(ServiceConfiguration config) throws IOException {
+        this.validationKey = getValidationKey(config);
+    }
+
+    @Override
+    public String getAuthMethodName() {
+        return "token";
+    }
+
+    @Override
+    public String authenticate(AuthenticationDataSource authData) throws AuthenticationException {
+        String token = null;
+
+        if (authData.hasDataFromCommand()) {
+            // Authenticate Pulsar binary connection
+            token = authData.getCommandData();
+        } else if (authData.hasDataFromHttp()) {
+            // Authentication HTTP request. The format here should be compliant to RFC-6750
+            // (https://tools.ietf.org/html/rfc6750#section-2.1). Eg:
+            //
+            // Authorization: Bearer xxxxxxxxxxxxx
+            String httpHeaderValue = authData.getHttpHeader(HTTP_HEADER_NAME);
+            if (!httpHeaderValue.startsWith(HTTP_HEADER_VALUE_PREFIX)) {
+                throw new AuthenticationException("Invalid HTTP Authorization header");
+            }
+
+            // Remove prefix
+            token = httpHeaderValue.substring(HTTP_HEADER_VALUE_PREFIX.length());
+        } else {
+            throw new AuthenticationException("No token credentials passed");
+        }
+
+        // Validate the token
+        try {
+            @SuppressWarnings("unchecked")
+            Jwt<?, Claims> jwt = Jwts.parser()
+                    .setSigningKey(validationKey)
+                    .parse(token);
+
+            return jwt.getBody().getSubject();
+        } catch (JwtException e) {
+            throw new AuthenticationException("Failed to authentication token: " + e.getMessage());
+        }
+    }
+
+    /**
+     * Try to get the validation key for tokens from several possible config options.
+     */
+    private static Key getValidationKey(ServiceConfiguration conf) throws IOException {
+        final boolean isPublicKey;
+        final String validationKeyConfig;
+
+        if (conf.getProperty(CONF_TOKEN_SECRET_KEY) != null) {
+            isPublicKey = false;
+            validationKeyConfig = (String) conf.getProperty(CONF_TOKEN_SECRET_KEY);
+        } else if (conf.getProperty(CONF_TOKEN_PUBLIC_KEY) != null) {
+            isPublicKey = true;
+            validationKeyConfig = (String) conf.getProperty(CONF_TOKEN_PUBLIC_KEY);
+        } else {
+            throw new IOException("No secret key was provided for token authentication");
+        }
+
+        final String validationKey;
+
+        if (validationKeyConfig.startsWith("data:")) {
+            log.info("Reading token validation key from config file");
+            validationKey = validationKeyConfig.substring("data:".length());
```

Comment:
This might be valid for a super simple case, but it wouldn’t work for cases that has mime-type or base64 encoding.

Why didn’t you use the URL class?

Reply:
... because I've never used `URL.openStream()` before :) 

Though, one thing is that the key is expected to be already base64 encoded. 

```shell
$ bin/pulsar tokens create-secret-key
secret-key CBqtuReZe+8nKqtJBbmuPTIg6b4W1KjOS21pcJkTJ/8=
```

Then one would just have to store the key string somewhere.

Reply:
I’m ok with parsing the schemes here and having a limitation in its format for now, but I still think this data scheme parsing is incorrect. It always need a comma after the colon. 
https://tools.ietf.org/html/rfc2397#section-3

Also, base64 encoding is not mandatory. You can pass URL encoded string instead.

Reply:
@maskit I have updated to conform with `data:` standard url, though I had to add manual parsing since java `URL` class doesn't support it.

Reply:
Did you look into #1212 I commented on another line? The URL class I suggested is `org.apache.pulsar.client.api.url.URL`. The parser already exists in it.